lib/Tunnel.js

```diff
@@ -1,158 +1,148 @@
-    var self = this;
-    var opt = self._opt;
-
-    var params = {
-        responseType: 'json'
+/* eslint-disable consistent-return, no-underscore-dangle */
+
+const url = require('url');
+const { EventEmitter } = require('events');
+const axios = require('axios');
+const debug = require('debug')('localtunnel:client');
+
+const TunnelCluster = require('./TunnelCluster');
+
+module.exports = class Tunnel extends EventEmitter {
+  constructor(opts = {}) {
+    super(opts);
+    this.opts = opts;
+    this.closed = false;
+    if (!this.opts.host) this.opts.host = 'https://localtunnel.me';
+  }
+
+  // initialize connection
+  // callback with connection info
+  _init(cb) {
+    const opt = this.opts;
+
+    const params = {
+      responseType: 'json',
     };
 
-    var base_uri = opt.host + '/';
-
+    const baseUri = `${opt.host}/`;
     // optionally override the upstream server
-    var upstream = url.parse(opt.host);
-
+    const upstream = url.parse(opt.host);
     // no subdomain at first, maybe use requested domain
-    var assigned_domain = opt.subdomain;
-
+    const assignedDomain = opt.subdomain;
     // where to quest
-    var uri = base_uri + ((assigned_domain) ? assigned_domain : '?new');
-
-    (function get_url() {
-        axios.get(uri, params)
-        .then(function(res){
-            var body = res.data;
-            if (res.status !== 200) {
-                var err =  new Error((body && body.message) || 'localtunnel server returned an error, please try again');
-                return cb(err);
-            }
-            var port = body.port;
-            var host = upstream.hostname;
-            var max_conn = body.max_conn_count || 1;
-            cb(null, {
-                remote_host: upstream.hostname,
-                remote_port: body.port,
-                name: body.id,
-                url: body.url,
-                max_conn: max_conn
-            });
-        })
-        .catch(function(err){
-            // TODO (shtylman) don't print to stdout?
-            console.log('tunnel server offline: ' + err.message + ', retry 1s');
-            return setTimeout(get_url, 1000);
+    const uri = baseUri + (assignedDomain || '?new');
+
+    (function getUrl() {
+      axios
+        .get(uri, params)
+        .then(res => {
+          const body = res.data;
+          debug('got tunnel information', res.data);
+          if (res.status !== 200) {
+            const err = new Error(
+              (body && body.message) || 'localtunnel server returned an error, please try again'
+            );
+            return cb(err);
+          }
+          const maxConn = body.max_conn_count || 1;
+          cb(null, {
+            remote_host: upstream.hostname,
+            remote_ip: body.ip,
+            remote_port: body.port,
+            name: body.id,
+            url: body.url,
+            cachedUrl: body.cachedUrl,
+            max_conn: maxConn,
+            local_https: opt.https,
```

Comment:
What's the benefit of returning these options back to the callback here? The caller has these options already since they constructed the initial localtunnel instance.

Reply:
In `open(cb)` it gets passed to `_establish`, which passes these on to `TunnelCluster`, which uses them, hence the need. The `cb` in `open` is actually invoked without any of this, so they aren't actually returned to the caller.